Add postgres support with tests

spencerldixon opened the pull request:
Building on top of Zach's PR and attempting to get this gem working with tests so we can merge postgresql support. Not entirely sure what I'm doing here so I would appreciate any input!

dblock commented:
Start by rebasing this code to fix the conflicts.

dblock commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 * [#28](https://github.com/slack-ruby/slack-ruby-bot-server/pull/28): Use slack-ruby-danger gem - [@dblock](https://github.com/dblock).
 * [#31](https://github.com/slack-ruby/slack-ruby-bot-server/pull/31): Adds MONGODB_URI as environment variable for MongoDB - [@corprew](https://github.com/corprew).
+* [#38](https://github.com/slack-ruby/slack-ruby-bot-server/pull/38): Add Postgres support, reference #7. - [@zachfeldman](https://github.com/zachfeldman).
```
There's an extra period in here, and you can remove "reference #7." altogether.

dblock commented on README.md:
```diff
@@ -31,6 +31,8 @@ You can use the [sample application](sample_app) to bootstrap your project and s
 
 Install [MongoDB](https://www.mongodb.org/downloads), required to store teams. We would like your help with [supporting other databases](https://github.com/slack-ruby/slack-ruby-bot-server/issues/12).
 
+You can also use PostgreSQL. See the PostgreSQL example - ensure you've filled out postgresql.yml and have setup your database correctly (user with the right name and privileges according to the file).
```
Lets make this two sections, maybe `### MongoDB` and `### PostgreSQL` with instructions on how to use each. You can remove "We would like your help with [supporting other databases](https://github.com/slack-ruby/slack-ruby-bot-server/issues/12)."

dblock commented on database.yml:
```diff
@@ -0,0 +1,18 @@
+default: &default
```
This file seems misplaced. I'd expect it to be in something like `config` folder.

dblock commented on database.yml:
```diff
@@ -0,0 +1,18 @@
+development:
+  <<: *default
+  database: slack_ruby_bot_server_development
+  username: slack_ruby_bot_server
```
I don't think this is how PostgreSQL default config works, we want this code to be able to run PostgreSQL tests without any changes.

dblock commented on lib/slack-ruby-bot-server.rb:
```diff
@@ -1,15 +1,22 @@
 require 'slack-ruby-bot-server/info'
-require 'slack-ruby-bot-server/models'
+
+if SlackRubyBotServer::Config.mongo?
```
Create two files, maybe `config/databases/mongo.rb` and `config/databases/postgresql.rb` and move all this logic there. The require should be something like `require 'config/databases/SlackRubyBotServer::Config.database_type`.

dblock commented on lib/slack-ruby-bot-server/api/presenters/team_presenter.rb:
```diff
@@ -10,7 +10,7 @@ module TeamPresenter
         property :name, type: String, desc: 'Team name.'
         property :domain, type: String, desc: 'Team domain.'
-        property :active, type: Boolean, desc: 'Team is active.'
+        property :active, type: 'Boolean', desc: 'Team is active.'
```
This shouldn't be changed.

dblock commented on lib/slack-ruby-bot-server/app.rb:
```diff
@@ -26,28 +25,47 @@ def logger
-    rescue Exception => e
-      warn "Error connecting to MongoDB: #{e.message}"
-      raise e
+      if SlackRubyBotServer::Config.mongo?
```
This should be moved to a database-specific class somewhere else IMO.

dblock commented on lib/slack-ruby-bot-server/config.rb:
```diff
@@ -3,11 +3,24 @@ module Config
     extend self
 
     attr_accessor :server_class
+    attr_accessor :database
```
Probably `database_type` is more appropriate. Even better, since we're talking about the adapter to access a database, maybe even `database_driver`?

dblock commented on lib/slack-ruby-bot-server/config.rb:
```diff
@@ -3,11 +3,24 @@ module Config
     end
 
+    def database
+      database ||= File.file?('config/mongoid.yml') ? :mongo : :postgresql
```
I think the right way to check is to look for `defined?(::Mongoid)` and such, also should fail if neither Mongoid or PostgreSQL are available.

dblock commented on sample_apps/sample_app_mongo/Gemfile:
```diff
@@ -1,8 +1,12 @@
 source 'https://rubygems.org'
```
mongo => mongoid

dblock commented on spec/models.rb:
```diff
@@ -0,0 +1,2 @@
+class Team < ActiveRecord::Base
```
This shouldn't be needed.

dblock commented on spec/schema.rb:
```diff
@@ -0,0 +1,15 @@
+ActiveRecord::Schema.define do
```
This should be in a pg-related subfolder.

dblock commented on spec/spec_helper.rb:
```diff
@@ -2,14 +2,35 @@
+files       = Dir[File.join(__dir__, 'support', '**/*.rb')]
+mongo_files = ["#{File.dirname(__FILE__)}/support/mongoid.rb"]
+
+if defined? Mongo
```
Spearate into subfolders and load, eg. "spec/support/databases/mongoid.rb" and "pg.rb"

dblock commented on spec/support/database_cleaner.rb:
```diff
@@ -7,7 +7,7 @@
 
   config.after :suite do
-    Mongoid.purge!
+    Mongoid.purge! if defined?(Mongo)
```
Move into the support folder's per-database file.

dblock commented:
Most importantly we need a build here that keys of DATABASE_ADAPTER=mongoid or DATABASE_ADAPTER=pgsql or similar in .travis.yml so that we can make sure all tests pass against one and the other.

dblock commented:
Also please see comments in https://github.com/slack-ruby/slack-ruby-bot-server/pull/38, a lot of the changes I'd like are still relevant here.

spencerldixon commented:
Thanks for the feedback @dblock, I'll get on it. Still not entirely sure how everything works but it's really interesting figuring it out, so please bare with me. One more thing, I'd like to use this in a Rails app I'm working on, would that be possible? Perhaps we could include some better instructions for doing this too

dblock commented:
Lets table the whole Rails thing for later, it's a whole other ballgame. Generally I would just use slack-ruby-bot in a Rails app.

spencerldixon commented:
Brilliant, I'm keen to work on rails integration too, I'm trying to create a bot app with rails that will need to support a personal assistant bot for individual members across multiple teams. Perhaps we can open an issue to start discussing this too, I know scale will also be an issue. Anyway, I've made a few changes, let me know your thoughts so far and I'll continue working on this

dblock commented on .travis.yml:
```diff
@@ -21,3 +21,6 @@ matrix:
   - "sh -e /etc/init.d/xvfb start"
+
+env:
+  - DB=postgres
```
Since there's a `matrix` above, you should modify the matrix to include 1 of each, something like

```
matrix:
   include:
     - rvm: 2.3.1
       script:
         - bundle exec danger
     - rvm: 2.3.1
        env:
          - DATABASE_ADAPTER=pg
     - rvm: 2.3.1
        env:
          - DATABASE_ADAPTER=mongoid
```

For now feel free to just remove ruby-head and jruby-head.

dblock commented on README.md:
```diff
@@ -29,7 +29,15 @@ Once a bot is registered, you can invite to a channel with `/invite @slackbotser
-Install [MongoDB](https://www.mongodb.org/downloads), required to store teams. We would like your help with [supporting other databases](https://github.com/slack-ruby/slack-ruby-bot-server/issues/12).
+### MongoDB
+
+Install [MongoDB](https://www.mongodb.org/downloads), required to store teams.
```
Add a note to use `mongoid`.

dblock commented on README.md:
```diff
@@ -29,7 +29,15 @@ Once a bot is registered, you can invite to a channel with `/invite @slackbotser
+
+### PostgreSQL
+
+You can also use PostgreSQL. See the PostgreSQL example - ensure you've filled out postgresql.yml and have setup your database correctly (user with the right name and privileges according to the file).
```
I would just remove "You can also use PostgreSQL" and make it similar to the mongoid instructions in every way.

dblock commented on config/databases/mongo.rb:
```diff
@@ -0,0 +1,3 @@
+require 'slack-ruby-bot-server/models/team/mongo.rb'
```
Just for consistency sake it's easier if the ENV variable keys the names, if `DATABASE_ADAPTER=mongoid` then this file should be called `database_adapters/mongoid.rb`.

spencerldixon commented:
Ok getting closer, I've moved everything to `database_adapters` folders and renamed the main config methods from postgresql to pg and mongo to mongoid to better reflect this. Let me know if I've missed anything though. Not sure how to go about hooking up the `DATABASE_ADAPTER` env in `.travis.yml` though to get this to run against the tests?

dblock commented on lib/slack-ruby-bot-server/config.rb:
```diff
@@ -11,20 +11,20 @@ def reset!
 
     def database_adapter
       database_adapter = if defined?(::Mongoid)
```
This should be `@database_adapter ||= ...`.

dblock commented on spec/spec_helper.rb:
```diff
@@ -11,8 +11,8 @@
-elsif SlackRubyBotServer::Config.mongo?
-  require 'spec/support/databases/mongo'
+if SlackRubyBotServer::Config.pg?
+  require 'spec/support/database_adapters/pg/*.rb'
```
I didn't know you could do a require with `*.rb`, learning every day :) 

This isn't a big deal, but it makes the order of requires unpredictable inside that folder. I would `require .../pg.rb` and inside `pg.rb` require anything that it may need.

dblock commented:
Getting really close! For rubocop, run `rubocop -a`, then `rubocop --auto-gen-config`.

dblock commented on lib/slack-ruby-bot-server/app.rb:
```diff
@@ -2,7 +2,6 @@ module SlackRubyBotServer
   class App
```
Should break this up and out into database-specific classes. For example you could have an `Adapter` class with a common interface, and `SlackRubyBotServeR::Config.database_adapter` would no longer be an enum, but an actual instance of that thing that you can call methods like `check!` and `prepare!`.

dblock commented on lib/slack-ruby-bot-server/config.rb:
```diff
@@ -3,11 +3,30 @@ module Config
       self.server_class = SlackRubyBotServer::Server
     end
 
+    def database_adapter
```
You could move this into an initializer. We call `reset!`, so maybe just set it there?

dblock commented:
Made some progress from here in https://github.com/slack-ruby/slack-ruby-bot-server/pull/48, closing this one. Still some test failures remaining.

dblock commented:
Released AR support in 0.6.0.
